File: cmd/code/spellchecker.go
package code

import "strings"

// Helper: check if rune is a vowel
func isVowel (ch rune) bool {
	return  ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u'
}

// Transform a word into its vowel-error key
func devowel(word string) string {
	var sb strings.Builder

	for _, ch:= range word {
		if isVowel(ch){
			sb.WriteRune('*')
		} else {
			sb.WriteRune(ch)
		}
	}

	return  sb.String()
}

// final function

func SpellChecker(wordlist []string, queries []string) []string {
	// exact match set
	exactWords := make(map[string] bool)

	// case insensitive match
	caseInsensitive := make(map[string] string)

	// vowel error map
	vowelInsensitive := make(map[string] string)

	for _, word := range wordlist {
		exactWords[word] = true

		lower := strings.ToLower(word)
		if _, exists := caseInsensitive[lower]; !exists {
			caseInsensitive[lower] = word
		} 

		dev := devowel(lower)

		if _, exists := vowelInsensitive[dev]; !exists {
			vowelInsensitive[dev] = word
		}
	}

	result := make([]string, len(queries))
	for i, q := range queries {
		// exact match
		if exactWords[q] {
			result[i] = q
			continue
		}

		lower := strings.ToLower(q)

		// case insensitive
		if val, exists := caseInsensitive[lower]; exists {
			result[i] = val
			continue
		} 

		// vowel insensitive
		dev := devowel(lower)
		if val, exists := vowelInsensitive[dev]; exists {
			result[i] = val
			continue
		}

		// no match
		result[i] = ""
	}

	return result
}

File: cmd/code/spellChecker.js

function spellChecker (wordlist, queries)
{
    // helper to check vowels
    const vowels = new Set( [ 'a', 'e', 'i', 'o', 'u' ] )
    const devowel = ( word ) => {
        word.toLowerCase().replace( /[aeiou]/g, '*' );
    }


    // exact word set
    const exactWords = new Set( wordlist )
    
    // case insensitive map
    const caseInsensitive = new Map()

    // vowel error map
    const vowelInsensitive = new Map()

    for ( const word of wordlist )
    {
        const lower = word.toLowerCase()
        if ( !caseInsensitive.has( lower ) )
        {
            caseInsensitive.set(lower, word)
        }

        const dev = devowel( word );
        if ( !vowelInsensitive.has( dev ) )
        {
            vowelInsensitive.set(dev, word)
        }

    }

    const result = []
    for ( const q of queries )
    {
        // exact match
        if ( exactWords.has( q ) )
        {
            result.push( q )
            continue
        }

        const lower = q.toLowerCase()

        // case insensitive
        if ( caseInsensitive.has( lower ) )
        {
            result.push( caseInsensitive.get( lower ) )
            continue
        }

        // vowel insensitive
        const dev = devowel( q )
        if ( vowelInsensitive.has( dev ) )
        {
            result.push( vowelInsensitive.get( dev ) )
            continue
        }

        // no match
        result.push("")
    }

    return result
}

const wordlist1 = ["KiTe","kite","hare","Hare"];
const queries1 = ["kite","Kite","KiTe","Hare","HARE","Hear","hear","keti","keet","keto"];
console.log(spellChecker(wordlist1, queries1));


const wordlist2 = ["yellowfds","aeiouasd", "AEUIsadOI", "fdsf", "aesdfjkg"];
const queries2 = ["YellOw", "aeiou", "AEUIOI"];
console.log(spellChecker(wordlist2, queries2));